WalletWasabi/Wallets/CpfpInfoProvider.cs

```diff
@@ -0,0 +1,201 @@
+using WalletWasabi.Models;
+using WalletWasabi.Tor.Http;
+using WalletWasabi.Tor.Http.Extensions;
+using WalletWasabi.WebClients.Wasabi;
+
+namespace WalletWasabi.Wallets;
+
+public class CpfpInfoProvider : BackgroundService
+{
+	private const int MaximumDelayInSeconds = 120;
+	private const int MaximumRequestsInParallel = 3;
+	private static readonly TimeSpan MinimumBetweenUpdateRequests = TimeSpan.FromMinutes(2);
+
+	public CpfpInfoProvider(WasabiHttpClientFactory httpClientFactory, Network network)
+	{
+		if (network == Network.Main)
+		{
+			HttpClient = httpClientFactory.NewHttpClient(() => new Uri("https://mempool.space/api/"), Tor.Socks5.Pool.Circuits.Mode.NewCircuitPerRequest);
+		}
+		else if(network == Network.TestNet)
+		{
+			HttpClient = httpClientFactory.NewHttpClient(() => new Uri("https://mempool.space/testnet/api/"), Tor.Socks5.Pool.Circuits.Mode.NewCircuitPerRequest);
+		}
+		else
+		{
+			throw new InvalidOperationException("CpfpInfoProvider is only operational on Main or TestNet");
+		}
+	}
+
+	public event EventHandler<EventArgs>? RequestedCpfpInfoArrived;
+
+	private ConcurrentDictionary<uint256, CachedCpfpInfo> CpfpInfoCache { get; } = new();
+
+	private IHttpClient HttpClient { get; }
+
+	private Channel<SmartTransaction> Channel { get; } = System.Threading.Channels.Channel.CreateUnbounded<SmartTransaction>();
+
+	private DateTime LastUpdateRequest { get; set; } = DateTime.MinValue;
+	private List<uint256> UpdateRequested { get; } = [];
+
+	private async Task FetchCpfpInfoAsync(SmartTransaction transaction, CancellationToken cancellationToken)
+	{
+		var txid = transaction.GetHash();
+
+		try
+		{
+			using CancellationTokenSource timeoutCts = new(TimeSpan.FromSeconds(60));
+			using CancellationTokenSource linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, timeoutCts.Token);
+
+			var response = await HttpClient.SendAsync(
+				HttpMethod.Get,
+				$"v1/cpfp/{txid}",
+				null,
+				linkedCts.Token).ConfigureAwait(false);
+
+			if (response.StatusCode != HttpStatusCode.OK)
+			{
+				await response.ThrowRequestExceptionFromContentAsync(cancellationToken).ConfigureAwait(false);
+			}
+
+			var cpfpInfo = await response.Content.ReadAsJsonAsync<CpfpInfo>().ConfigureAwait(false);
+
+			CpfpInfoCache.AddOrReplace(txid, new CachedCpfpInfo(cpfpInfo, transaction));
+
+			RequestedCpfpInfoArrived?.Invoke(this, EventArgs.Empty);
+		}
+		catch (OperationCanceledException)
+		{
+			Logger.LogTrace("Request was cancelled by exiting the app.");
+		}
+	}
+
+	private bool ShouldRequest(SmartTransaction tx, bool requestIfForeignOutput, bool ignoreCache)
+	{
+		if (tx.Confirmed)
+		{
+			// Don't request if the transaction is already confirmed.
+			return false;
+		}
+		if (tx.ForeignInputs.Count == 0 &&
+		    (!requestIfForeignOutput || tx.ForeignOutputs.Count == 0))
+		{
+			// Don't request if there are no foreign inputs and whether no foreign output whether we don't want to request in that case.
```

Comment:
Maybe this makes more sense.

```suggestion
			// Don't request if there are no foreign inputs and either no foreign outputs or in case we don't want to request.
```